src/main/java/org/jenkinsci/plugins/workflow/libs/LibraryRecord.java

```diff
@@ -54,15 +61,33 @@ public final class LibraryRecord {
      * @param trusted Whether the library is trusted. Typically determined by {@link LibraryResolver#isTrusted}, but see also {@link LibraryStep}.
      * @param changelog Whether we should include any SCM changes in this library in the build's changelog.
      * @param cachingConfiguration If non-null, contains cache-related configuration.
+     * @param libraryPath The path to the library within the repository. Used in {@link SCMBasedRetriever}. If null, the library is assumed to be at the root of the repository.
      * @param source A string describing the source of the configuration of this library. Typically the class name of a {@link LibraryResolver}, sometimes with additional data, but see also {@link LibraryStep}.
      */
-    LibraryRecord(String name, String version, boolean trusted, boolean changelog, LibraryCachingConfiguration cachingConfiguration, String source) {
+    LibraryRecord(String name, String version, boolean trusted, boolean changelog, LibraryCachingConfiguration cachingConfiguration, String source, @CheckForNull String libraryPath) {
         this.name = name;
         this.version = version;
         this.trusted = trusted;
         this.changelog = changelog;
         this.cachingConfiguration = cachingConfiguration;
-        this.directoryName = directoryNameFor(name, version, String.valueOf(trusted), source);
+        logString = this.name + "@" + this.version;
+        if (onTheRoadToNowhere(libraryPath)) {
+            this.libraryPath = "";
+            this.directoryName = directoryNameFor(name, version, String.valueOf(trusted), source);
+        } else {
+            this.libraryPath = libraryPath;
+            this.directoryName = directoryNameFor(name, version, String.valueOf(trusted), source, libraryPath);
+            logString += " from libraryPath: " + libraryPath;
```

Comment:
```suggestion
            logString += ":" + libraryPath;
```
Might look slightly nicer?

Reply:
Done